V2

jxnblk opened the pull request:
# v2

This is presented as a prototype for discussion. This doesn't have to be the actual v2, but I do like a lot of the things happening here. Would love some feedback and questions.

### What

- Smaller, simplified, and reorganized object
- Reduce duplication with PostCSS functionality
- Works as a postcss plugin
- Remove stats that can be determined in the app/presentation layer
- Lite option
- Full property names are used as keys rather than camel casing
- Uses [standard](https://www.npmjs.com/package/standard)
- **New** Helper methods

### New returned object

```js
// n = number
// str = string
// [] = array
// prop = CSS property key

{
  size: n,
  gzipSize: n,
  rules: {
    total: n,
    size: {
      graph: [n], // Array of numbers for each rule's size (e.g. number of declarations)
      max: n,
      average: n
    }
  },
  selectors: {
    total: n,
    id: n,              // total number of id selectors
    class: n,           // total number of class selectors
    type: n,            // total number of type selectors (not currently implemented)
    pseudoClass: n,     // total number of pseudo class selectors
    psuedoElement: n,   // total number of pseudo element selectors
    repeated: [str],    // array of repeated selectors
    values: [str],      // array of all selectors
    specificity: {
      max: n
      average: n
    },
    getSpecificityGraph(),
    getRepeatedSelectors(),
    getSortedSpecificity()
  },
  declarations: {
    total: n,
    important: n,       // total number of important declarations
    vendorPrefix: n,    // total number of vendor prefixed declarations
    properties:         // object for each unique property
      prop: [str]       // array of values for each property
    },
    getPropertyResets(),
    getUniquePropertyCount(),
    getPropertyValueCount()
  },
  mediaQueries: {
    total: n,
    unique: n,
    values: [str]       // array of values for media queries (e.g. '(min-width: 40em)')
  }
}
```

### To do

- [x] Implement lite option
- [x] Clean up unused modules and/or create new modules
- [x] Count type selectors
- [x] Update README
- [x] Account for missing _reset_ declarations
- [x] Update/add new tests

### Questions

- Does this organizational scheme make sense?
- Can/should we handle any shorthand property parsing at the presentation layer?
- Is there anything else we'd want to parse for media queries?
- How should the lite option be implemented? Should `opts` be passed to each module in `lib`? Or should the modules be split up further?

### Usage examples

**UPDATE**
I've added some helper methods that make the examples below redundant.

#### Get total number of unique colors

```js
var _ = require('lodash')
var cssstats = require('cssstats')

var stats = cssstats(css)

var uniqueColorsCount = _.uniq(stats.declarations.properties.colors).length
```

#### `display: none` count

```js
var displayNoneCount = stats.declarations.properties.display
  .filter(function (val) {
    return val === 'none'
  })
  .length
```

#### Find the selector with the highest specificity

```js
var maxSpecificity = _.max(stats.selectors.specificity.graph)
var index = stats.selectors.specificity.graph.indexOf(maxSpecificity)
var maxSelector = stats.selectors.values[index]
```

#### Sort selectors by highest specificity

```js
var ranked = _.zipWith(stats.selectors.values, stats.selectors.specificity.graph, function (a, b) {
    return {
      selector: a,
      specificity: b
    }
  })
  .sort(function (a, b) {
    return b.specificity - a.specificity
  })
```

#### Specificity graph

```js
var graph = stats.selectors.specificity.graph
```

#### Rule size graph

```js
var graph = stats.rules.size.graph
```

HipsterBrown commented:
Can you clarify what you mean by "lite option"?

Could organizing the returned object's properties alphabetically be the simplest convention anyone reading through it?

Besides those questions, this looks like a rad update. I love the use case examples and performance optimizations overall. I've rarely seen a better looking PR. :heart:

jxnblk commented:
@HipsterBrown 
- `lite` option would mean arbitrarily removing some of the larger objects and arrays (e.g. `declarations.properties`) to help with keeping the size down and for performance.
- Sorting the properties object could make sense. Currently it's sorted by source order, which I tend to prefer, but might not matter as much for this.
- Thanks! :sunglasses:

johnotander commented on lib/selectors.js:
```diff
@@ -1,78 +1,86 @@
+    result.specificity.graph = graph
+    result.specificity.max = _.max(graph)
+    result.specificity.average = _.sum(graph) / graph.length
+    result.repeated = _.uniq(
```
:heart_eyes_cat:

rafaelrinaldi commented:
:clap: This is beautiful! I was thinking of improving the return object while working on a project that relies  on `css-statistics`.

Just one thing though, what if properties like `vendorPrefix` or `important` return an array of occurences instead of a number only? This way its easier to spot each one shows up the most and things like that.

johnotander commented:
This is amazing! Loving how this approach is cleaning up the code so drastically while removing a lot of duplicated return content.

### tl;dr

:shipit: 

***

> Does this organizational scheme make sense?

Definitely.

> Can/should we handle any shorthand property parsing at the presentation layer?

I've been pondering this one a bit, and I think that the presentation layer is likely the best location for this. It feels kinda hacky to add a `font-size: 123px` declaration in the returned object when it already exists in the `font` shorthand.

So, it's likely best to expose a function that is contextually aware of something like `font-size` and the `font` shorthand. How exactly is the best way to do this, I'm not quite sure.

> Is there anything else we'd want to parse for media queries?

I think it _may_ be useful to expose some stats on the CSS contained within each media query, as this could help illustrate CSS with architectural problems. However, this could be handled in the usage section similar to the other examples you've outlined.

Also, this new approach will take care of cssstats/css-statistics/pull/12! :dancers: 

> How should the lite option be implemented? Should opts be passed to each module in lib? Or should the modules be split up further?

IMO, the libs should receive the option and handle it accordingly. I think we could also take further steps to modularize this project by breaking a lot of the lib stuff into standalone modules. We should also break the cli into `cssstats-cli`, allowing us to add more CLI functionality without adding complexity directly to this core module.

> what if properties like vendorPrefix or important return an array of occurences 

I agree with @rafaelrinaldi here, that would be great!

#### Usage examples

The usage example section is wonderful, too. I like the idea of not bloating the returned object with data that is less likely to be used. Though, I think we should probably either expose these as options, i.e. `{ displayNone: true }` or another module like `cssstats-display-none`.

## A pluggable API?

I think it might be worthwhile considering some type of pluggable API. That way users could incorporate their own functions and require more esoteric stats examples if they'd like:

```js
var cssstats = require('cssstats')

cssstats(myCss).use(require('cssstats-display-none'), require('cssstats-specificity-sort'), function(stats) {
  // ... Do some crazy stuff with the stats object
})
```

Not exactly sure how this would work, but somehow each of these "plugins" of additional statistical information could somehow add to the returned stats object.

Thoughts?

jxnblk commented:
This is super helpful. Thanks!

@rafaelrinaldi @johnotander as far as the `vendorPrefix` and `important` arrays go, what should they contain? Both the property and value for each declaration? 

My thinking with the shorthand properties and other example is that they might make sense as methods, such as `stats.getAllFontSizes()`. Any thoughts on that?

I'm down to add more information to the media query object – any ideas on how to parse the contents of a media query block and what to show?

Passing `opts` into the modules seems like an okay solution. An example of the other idea would be to have both a `lib/selectors` and a `lib/selectors-specificity` module. Things like this could also just be methods that are returned with the object.

The idea of a pluggable API is interesting. We could probably rely on the postcss plugin architecture for that – there's probably a sane way to tack on any postcss plugin to cssstats.

johnotander commented:
> as far as the vendorPrefix and important arrays go, what should they contain? Both the property and value for each declaration?

I'd say the property and value. As it'd really shed some light on any duplicated declarations with the `!important` keyword.

> My thinking with the shorthand properties and other example is that they might make sense as methods, such as stats.getAllFontSizes(). Any thoughts on that?

:+1:

> I'm down to add more information to the media query object – any ideas on how to parse the contents of a media query block and what to show?

Off the top of my head, I believe all the contained styling is available as `.nodes` (or `.childNodes` maybe?) within the `atrule` object. We could then pass that sub-AST to `stats-lite` to get some really basic data.

> An example of the other idea would be to have both a lib/selectors and a lib/selectors-specificity module. Things like this could also just be methods that are returned with the object.

I think this might be the best approach. Though, we will then have to consider how we convert the returned object to JSON. Perhaps exposing a `.toJSON()` method? Also, this _could_ potentially be solved if we run with some sort of plugin mechanism, too. Please see below.

> We could probably rely on the postcss plugin architecture for that – there's probably a sane way to tack on any postcss plugin to cssstats.

Yeah, I'm not quite sure if this is something that would be worthwhile/useful to end users. Especially if it's not easy to somehow lean on PostCSS to solve this (I will have to do some digging).

Ultimately, the main problem I'm seeing here is that we have different levels/scopes of information that we want to expose in terms of data/stats: Lite, Default, All The Things, Custom Things. It might just be as simple as making options: `lite`, `all`, `displayNone`, etc. However, the drawback I see there is the configuration could be a little unwieldy (but likely not more unwieldy than calling a plugin). Actually, as I type this up, I wonder if this is the best route rather than implementing some type of plugin infrastructure. We can expose a slew of options, with sensible defaults, to configure the statistical information that we return.

jxnblk commented:
On a side note, I would *love* to rename this repo `cssstats-core` if there's no opposition.

jxnblk commented on lib/media-queries.js:
```diff
@@ -0,0 +1,46 @@
+    total: 0,
+    unique: 0,
+    values: [],
+    contents: {} // Not sure what to name this
```
Any ideas on what to call this? Essentially it's stats for each media query.

jxnblk commented on lib/media-queries.js:
```diff
@@ -0,0 +1,46 @@
+      stats.rules = rules(qRoot, opts)
+      stats.selectors = selectors(qRoot, opts)
+      stats.declarations = declarations(qRoot, opts)
+      result.contents[key] = stats
```
It might be better to just push media query contents to an array and handle combining similar ones outside of the core module.

johnotander replied:
Yeah, the per media-query stats perhaps don't need to be in the core. Though, the node contents will need to be available so that it has the necessary data to further analyze.

mrmrs commented:
This is epic and awesome. 

No qualms from me on renaming the module. 

These examples are fire. A+ work there.

For Media Queries - don't know if this should be handled by the returned object here but I would love to see:

- boolean for whether something other than em is being used
- boolean for whether or not max-width architecture is present i.e not a range of min-width / max-width
but just max-width by itself. 
- Filtering stats by Media Query. e.g how many total / unique font-sizes per media query.

jxnblk commented:
@mrmrs so it *should* be returning stats for each media query block now, which means you could find the total/unique properties per media query like the root example.

As for the ems, and max/min-width, we'd need something to parse the media query expressions (i.e. values/params), since postcss only returns a string for those.

jxnblk commented:
Also was just quickly experimenting with adding some convenience methods in, and this already feels like a really good API.

Examples:
- `stats.getSpecificityGraph()`
- `stats.getPropertyUniques('color')`
- `stats.getPropertyValueCount('display', 'none')
- `stats.getHighestSpecificitySelectors()`

Adding a `.toJSON` method to the object should prevent these from being stringified from what I can tell.

mrmrs commented:
@jxnblk - fuck yes. That is awesome.

rafaelrinaldi commented:
@jxnblk Well, using the vendor prefix example we could have something like this:

```
{
  vendorPrefixes: {
    '-ms-border-radius': 3,
    '-ms-transition-delay': 7
  }
}
```
Here we can easily tell how many vendor prefixes have been found and how many times they appear.

jxnblk commented:
Relevant to the media query discussion: https://github.com/cvrebert/css-mq-parser

johnotander commented on lib/media-queries.js:
```diff
@@ -0,0 +1,43 @@
+    total: 0,
+    unique: 0,
+    values: [],
+    contents: [] // Not sure what to name this
```
I wonder if this should be keyed off the values? That way the `contents` are associated with the media query value.

```js
var result = {
  total: 0,
  unique: 0,
  values: [],
  statsByValue: {
    '(min-width: 40em)': {}
  }
}
```

jxnblk replied:
Yeah, I originally had it as an object, but thought it might be useful to have each block listed separately. Mostly just exploring how to handle the media queries at this point.

johnotander replied:
Also, with this proposed implementation, are media query stats also making into the overall stats for the CSS? I'm failing to see where that is occurring currently.

jxnblk replied:
Doesn't `postcss.eachRule` walk through the media query blocks as well?

johnotander replied:
Ah, yes. You are [correct](https://github.com/postcss/postcss/blob/master/docs/api.md#containereachrulecallback).

johnotander commented:
Based off your proposed PR #43 which adds the helper methods, we could add options to configure what the returned object includes/excludes. This would be a lot simpler than a pluggable API, but still allows for thorough customization/optimization for environments where those concerns are present.

We could propose defaults, and modify them like so:

```js
postcss
  .use(cssstats({
    specificityGraph: false,
    propertyResets: true
  })
  .process(myCss)
```

jxnblk commented:
@johnotander I was thinking of something similar. Is there any harm in always including the methods by default? I was imagining that options could be added to do something like this:

```js
if (opts.propertyResets) {
  result.resets = result.getPropertyResets()
}
```

I guess with what you're proposing, it would help when using the module in client-side applications.

johnotander commented:
> Is there any harm in always including the methods by default?

I don't think there's any harm, so that is probably the way to go.

> options could be added to do something like this [js code]

:+1:

jxnblk commented:
I've added helper methods in, as well as options to include the results of relevant methods in the object for stringifying.

@rafaelrinaldi @johnotander For important declarations, they now have an option `importantDeclarations` that will add an array of `!important` declarations with properties and values. For vendor prefixes, `declarations.getVendorPrefixed()` will return an array of declarations. Setting the `vendorPrefixedProperties` option will include that array in the results.

jxnblk commented:
This pull request fixes #2 and is related to:
- cssstats/cssstats#163
- cssstats/cssstats#147
- cssstats/cssstats#146
- cssstats/cssstats#132
- cssstats/cssstats#126
- cssstats/cssstats#102
- cssstats/cssstats#22
- cssstats/cssstats#21
- cssstats/cssstats#15

jxnblk commented:
Just published a beta if anyone ones to test it out: `npm i cssstats@beta`

johnotander commented:
Just did some hand testing with the beta, comparing it to the output with cssstats.com and it looks accurate. Not to mention, the returned object is _much_ more human readable. Awesome work!

:shipit:

jxnblk commented:
I've added and updated tests and have been trying this out in the new app. Everything feels good to me, and if there's no objections, I'd like to publish an official release today.

johnotander commented on README.md:
```diff
@@ -45,41 +56,187 @@ cat some-css-file.css | cssstats
-__`selectors`:__ An array of selectors sorted by source order with the selector string, specificity score, and parts array
+- `safe` (boolean, default: `true`) - enables PostCSS safe mode for parsing CSS with syntax errors
+- `mediaQueries` (boolean, default `true`) - determines whether or not to generate stats for each media query block
+- `importantDeclarations` (boolean, deault `false`) - include an array of declarations with `!important`
```
default

johnotander commented on README.md:
```diff
@@ -45,41 +56,187 @@ cat some-css-file.css | cssstats
+#### `rules` object
+
+- `total` - total number of rules
+- `size` object
```
- object

jxnblk replied:
Any advice on a standard way of writing this out would be cool. I don't know if using something like jsdoc is still cool.

johnotander replied:
I like the current formatting. Though, I'd probably prefer to see the type for all properties: `number`, `object`, `array`, etc. IMO it's pretty useful when scanning the API docs.

jxnblk replied:
Good call, I'll add those in. Related: does that weird fence block with the returned object make sense or is there a better way to represent that?

johnotander replied:
Made sense to me.

johnotander commented:
> I'd like to publish an official release today.

:dancers: :dancers: :dancers:
